### PR TITLE
feat: add advanced logger with exception support

### PR DESCRIPTION
**Pull request by gabrielfs7**
https://oat-sa.atlassian.net/browse/ADF-840

- [ ] https://github.com/oat-sa/tao-core/pull/3243

### CODE REVIEW

**Review by hectoras — APPROVED** [review]

File: common/oatbox/log/logger/extender/RequestContextExtender.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        return $this;
+    }
+
+    public function extend(array $context): array
+    {
+        $context[self::CONTEXT_REQUEST_DATA] = isset($context[self::CONTEXT_REQUEST_DATA])
+        && is_array($context[self::CONTEXT_REQUEST_DATA])
+            ? array_merge($context[self::CONTEXT_REQUEST_DATA], $this->getContextRequestData())
+            : $this->getContextRequestData();
```
Comment (shpran) [inline common/oatbox/log/logger/extender/RequestContextExtender.php:45]:
```suggestion
        $context[self::CONTEXT_REQUEST_DATA] = isset($context[self::CONTEXT_REQUEST_DATA])
            && is_array($context[self::CONTEXT_REQUEST_DATA])
                ? array_merge($context[self::CONTEXT_REQUEST_DATA], $this->getContextRequestData())
                : $this->getContextRequestData();
```

File: common/oatbox/log/logger/extender/UserContextExtender.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        $this->sessionService = $sessionService;
+    }
+
+    public function extend(array $context): array
+    {
+        $context[self::CONTEXT_USER_DATA] = isset($context[self::CONTEXT_USER_DATA])
+            && is_array($context[self::CONTEXT_USER_DATA])
+            ? array_merge($context[self::CONTEXT_USER_DATA], $this->getContextUserData())
+            : $this->getContextUserData();
```
Comment (shpran) [inline common/oatbox/log/logger/extender/UserContextExtender.php:46]:
```suggestion
        $context[self::CONTEXT_USER_DATA] = isset($context[self::CONTEXT_USER_DATA])
            && is_array($context[self::CONTEXT_USER_DATA])
                ? array_merge($context[self::CONTEXT_USER_DATA], $this->getContextUserData())
                : $this->getContextUserData();
```

File: common/oatbox/log/logger/extender/UserContextExtender.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    private function getUserIdentifier(): ?string
+    {
+        try {
+            if ($this->sessionService->isAnonymous()) {
+                return 'anonymous';
+            }
+
+            $user = $this->sessionService->getCurrentUser();
+
+            return $user ? $user->getIdentifier() : 'system';
```
Comment (shpran) [inline common/oatbox/log/logger/extender/UserContextExtender.php:71]: Wdyt?
In this case, we can change return type from `?string` to `string`
```suggestion
            return $user
                ? $user->getIdentifier() ?? 'noIdentifier'
                : 'system';
```

File: test/unit/common/oatbox/log/logger/AdvancedLoggerTest.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        $this->contextExtender
+            ->method('extend')
+            ->willReturn(
+                [
+                    'extended' => true,
+                ]
+            );
+
+        $this->logger
+            ->expects($this->exactly(1))
```
Comment (shpran) [inline test/unit/common/oatbox/log/logger/AdvancedLoggerTest.php:62]:
```suggestion
            ->expects($this->once())
```

File: test/unit/common/oatbox/log/logger/AdvancedLoggerTest.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+namespace oat\generis\test\unit\common\oatbox\log\logger;
```
Comment (shpran) [inline test/unit/common/oatbox/log/logger/AdvancedLoggerTest.php:21]:
```suggestion
 */

declare(strict_types=1);

namespace oat\generis\test\unit\common\oatbox\log\logger;
```

File: test/unit/common/oatbox/log/logger/extender/ExceptionContextExtenderTest.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+namespace oat\generis\test\unit\common\oatbox\log\logger\extender;
```
Comment (shpran) [inline test/unit/common/oatbox/log/logger/extender/ExceptionContextExtenderTest.php:21]:
```suggestion
 */

declare(strict_types=1);

namespace oat\generis\test\unit\common\oatbox\log\logger\extender;
```

File: test/unit/common/oatbox/log/logger/extender/RequestContextExtenderTest.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+namespace oat\generis\test\unit\common\oatbox\log\logger\extender;
```
Comment (shpran) [inline test/unit/common/oatbox/log/logger/extender/RequestContextExtenderTest.php:21]:
```suggestion
 */

declare(strict_types=1);

namespace oat\generis\test\unit\common\oatbox\log\logger\extender;
```

File: test/unit/common/oatbox/log/logger/extender/UserContextExtenderTest.php [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+namespace oat\generis\test\unit\common\oatbox\log\logger\extender;
```
Comment (shpran) [inline test/unit/common/oatbox/log/logger/extender/UserContextExtenderTest.php:21]:
```suggestion
 */

declare(strict_types=1);

namespace oat\generis\test\unit\common\oatbox\log\logger\extender;
```

File: common/oatbox/log/logger/extender/UserContextExtender.php (hunk 3 above)
Comment (gabrielfs7) [reply common/oatbox/log/logger/extender/UserContextExtender.php:77]: In this case, I would prefer so store null, cause it is an edge case.

**Review by shpran — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Review by hectoras — APPROVED** [review]

**Comment by AntonNeviarouski** [comment]
QA verifed
